src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs

```diff
@@ -10,19 +12,26 @@ internal partial class LazyItemEvaluator<P, I, M, D>
     {
         class RemoveOperation : LazyItemOperation
         {
+            readonly ImmutableList<string> _matchOnMetadata;
+
             public RemoveOperation(OperationBuilder builder, LazyItemEvaluator<P, I, M, D> lazyEvaluator)
                 : base(builder, lazyEvaluator)
             {
+                _matchOnMetadata = builder.ItemElement.MatchOnMetadata.Equals(string.Empty) ? null : builder.ItemElement.MatchOnMetadata.Split(';').ToImmutableList();
             }
 
             // todo port the self referencing matching optimization (e.g. <I Remove="@(I)">) from Update to Remove as well. Ideally make one mechanism for both. https://github.com/Microsoft/msbuild/issues/2314
             // todo Perf: do not match against the globs: https://github.com/Microsoft/msbuild/issues/2329
             protected override ImmutableList<I> SelectItems(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
             {
+                if (_matchOnMetadata != null && (_itemSpec.Fragments.Count != 1 || _itemSpec.ItemSpecString.Contains('%')))
+                {
+                    throw new InternalErrorException("Only a single item reference allowed when removing with metadata match");
+                }
                 var items = ImmutableHashSet.CreateBuilder<I>();
                 foreach (ItemData item in listBuilder)
                 {
-                    if (_itemSpec.MatchesItem(item.Item))
+                    if ((_matchOnMetadata == null && _itemSpec.MatchesItem(item.Item)) || (_matchOnMetadata != null && _itemSpec.MatchesItemOnMetadata(item.Item, _matchOnMetadata)))
```

Comment:
To split msbuild expressions by semicolon use `ExpressionShredder.SplitSemiColonSeparatedList`.

Add tests to ensure that `MatchOnMetadata` can contain references to properties and items (e.g. `A Remove="@(B)" MatchOnMetadata="$(P);@(C)"`)

Remove under targets should already expand properties and items, and that happens here: https://github.com/microsoft/msbuild/pull/4997/files#diff-0d4693ff9ac2fd56e6fda64d0fd0c6cbR108

For Remove not under targets, I think you also need to expand it somewhere, and do other lazy item evaluation bookkeeping. You'll have to follow the example for what Include does with its Exclude attribute:
- process `MatchOnMetadata` for the Remove operation similarly to how the Include operation processes its `Exclude`: https://github.com/microsoft/msbuild/blob/master/src/Build/Evaluation/LazyItemEvaluator.cs#L492-L509
- save the processed `MatchOnMetadata` results in a new RemoveOperationBuilder, which the gets passed into and processed by the RemoveOperation constructor.

Reply:
Oops, this comment is intended for this diff: https://github.com/microsoft/msbuild/pull/4997/commits/9738d93462e6551f2f7d30acce00bc28d81c5cf8#diff-79dabcabc9c2b9dc24b1cfcd09ee532cR20